src/users/amoc_users_worker_sup.erl

```diff
@@ -0,0 +1,174 @@
+%% @private
+%% @copyright 2024 Erlang Solutions Ltd.
+%% @doc Supervisor-like gen_server with some tracking responsibilities over the users
+%%
+%% We want to keep consistent track of all users running globally by running special code upon a
+%% children's death. Standard solutions don't cut it because:
+%%  - A supervisor doesn't expose callbacks on user termination
+%%  - Implementing code on the user process before it dies risks inconsistencies if it is killed
+%%  More improvements that could be made would be to distribute the supervision tree like ranch did,
+%%  see https://stressgrid.com/blog/100k_cps_with_elixir/
+%% @end
+-module(amoc_users_worker_sup).
+
+-behaviour(gen_server).
+
+%% gen_server callbacks
+-export([start_link/1]).
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2, terminate/2]).
+
+-export([start_child/4, stop_child/2, start_children/4, stop_children/3, terminate_all_children/1]).
+
+-export([get_all_children/1]).
+
+-record(state, {
+          index :: non_neg_integer(),
+          tid :: ets:tid(),
+          tasks = #{} :: #{reference() => pid()}
+         }).
+-type state() :: #state{}.
+
+-define(SHUTDOWN_TIMEOUT, 2000). %% 2 seconds
+
+%% @private
+-spec start_link(non_neg_integer()) -> {ok, pid()}.
+start_link(N) ->
+    gen_server:start_link(?MODULE, N, []).
+
+-spec start_child(pid(), amoc:scenario(), amoc_scenario:user_id(), any()) -> ok.
+start_child(Sup, Scenario, Id, ScenarioState) ->
+    gen_server:cast(Sup, {start_child, Scenario, Id, ScenarioState}).
+
+-spec start_children(pid(), amoc:scenario(), [amoc_scenario:user_id()], any()) -> ok.
+start_children(Sup, Scenario, UserIds, ScenarioState) ->
+    gen_server:cast(Sup, {start_children, Scenario, UserIds, ScenarioState}).
+
+-spec stop_children(pid(), non_neg_integer(), boolean()) -> ok.
+stop_children(Sup, Count, Force) ->
+    gen_server:cast(Sup, {stop_children, Count, Force}).
+
+-spec terminate_all_children(pid()) -> any().
+terminate_all_children(Sup) ->
+    gen_server:cast(Sup, terminate_all_children).
+
+-spec stop_child(pid(), boolean()) -> ok.
+stop_child(Pid, false) ->
+    exit(Pid, shutdown),
+    ok;
+stop_child(Pids, true) ->
+    spawn(shutdown_and_kill_after_timeout_fun(Pids)),
+    ok.
+
+-spec get_all_children(pid()) -> [{pid(), amoc_scenario:user_id()}].
+get_all_children(Sup) ->
+    gen_server:call(Sup, get_all_children, infinity).
+
+%% @private
+-spec init(non_neg_integer()) -> {ok, term()}.
+init(N) ->
+    process_flag(trap_exit, true),
+    Tid = ets:new(?MODULE,  [ordered_set, protected]),
```

Comment:
please make it named `list_to_atom(atom_to_list(__MODULE__) ++ integer_to_list(tid))`
so it's easy to identify the table.